File: quickstarts/VB/Stoplight/Src/UI/StoplightForm.vb
' Copyright (c) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.

Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports Microsoft.Practices.Unity

Namespace StopLight.UI
	Public Partial Class StoplightForm
		Inherits Form
		Implements IStoplightView
    Private _presenter As StoplightPresenter

    Public Sub New()
      InitializeComponent()
    End Sub

    <Dependency()> _
    Public Property Presenter() As StoplightPresenter
      Get
        Return _presenter
      End Get
      Set(ByVal value As StoplightPresenter)
        _presenter = value
        presenter.SetView(Me)
      End Set
    End Property

#Region "IStoplightView Members"

    Public Property CurrentColor() As Color Implements IStoplightView.CurrentColor
      Get
        Return stopLightPanel.BackColor
      End Get
      Set(ByVal value As Color)
        stopLightPanel.BackColor = value
        RaisePropertyChanged(StoplightViewProperties.CurrentColor)
      End Set
    End Property

    Public Property GreenDuration() As String Implements IStoplightView.GreenDuration
      Get
        Return greenDurationTextBox.Text
      End Get
      Set(ByVal value As String)
        greenDurationTextBox.Text = value
        RaisePropertyChanged(StoplightViewProperties.GreenDuration)
      End Set
    End Property

    Public Property YellowDuration() As String Implements IStoplightView.YellowDuration
      Get
        Return yellowDurationTextBox.Text
      End Get
      Set(ByVal value As String)
        yellowDurationTextBox.Text = value
        RaisePropertyChanged(StoplightViewProperties.YellowDuration)
      End Set
    End Property

    Public Property RedDuration() As String Implements IStoplightView.RedDuration
      Get
        Return redDurationTextBox.Text
      End Get
      Set(ByVal value As String)
        redDurationTextBox.Text = value
        RaisePropertyChanged(StoplightViewProperties.RedDuration)
      End Set
    End Property

    Public Event UpdateClicked As EventHandler Implements IStoplightView.UpdateClicked

    Public Event ForceChangeClicked As EventHandler Implements IStoplightView.ForceChangeClicked

    Public Sub SetError(ByVal propertyName As String, ByVal errorMessage As String) Implements IStoplightView.SetError
      Dim controlsByName As New Dictionary(Of String, Control)()
      controlsByName.Add(StoplightViewProperties.GreenDuration, greenDurationTextBox)
      controlsByName.Add(StoplightViewProperties.YellowDuration, yellowDurationTextBox)
      controlsByName.Add(StoplightViewProperties.RedDuration, redDurationTextBox)


      If controlsByName.ContainsKey(propertyName) Then
        errorProvider.SetError(controlsByName(propertyName), errorMessage)
      End If
    End Sub

#End Region

#Region "INotifyPropertyChanged Members"

    Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

#End Region

		' Event firing helpers

		Protected Overridable Sub RaisePropertyChanged(ByVal propertyName As String)
      Dim handlers As PropertyChangedEventHandler = PropertyChangedEvent
      If Not handlers Is Nothing Then
        handlers(Me, New PropertyChangedEventArgs(propertyName))
      End If
		End Sub

		Protected Overridable Sub RaiseUpdateClicked()
      Dim handlers As EventHandler = UpdateClickedEvent
      If Not handlers Is Nothing Then
        handlers(Me, EventArgs.Empty)
      End If
		End Sub

		Protected Overridable Sub RaiseForceChangeClicked()
      Dim handlers As EventHandler = ForceChangeClickedEvent
      If Not handlers Is Nothing Then
        handlers(Me, EventArgs.Empty)
      End If
		End Sub

		Private Sub OnUpdateScheduleClicked(ByVal sender As Object, ByVal e As EventArgs)
			RaiseUpdateClicked()
		End Sub

		Private Sub OnChangeLightClicked(ByVal sender As Object, ByVal e As EventArgs)
			RaiseForceChangeClicked()
		End Sub
	End Class
End Namespace

File: quickstarts/CS/Stoplight/Src/UI/StoplightForm.cs
// Copyright (c) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using Microsoft.Practices.Unity;

namespace StopLight.UI
{
	public partial class StoplightForm : Form, IStoplightView
	{
		private StoplightPresenter presenter;

		public StoplightForm()
		{
			InitializeComponent();
		}

		[Dependency]
		public StoplightPresenter Presenter
		{
			get { return presenter; }
			set
			{
				presenter = value;
				presenter.SetView(this);
			}
		}

		#region IStoplightView Members

		public Color CurrentColor
		{
			get { return stopLightPanel.BackColor; }
			set
			{
				stopLightPanel.BackColor = value;
				RaisePropertyChanged(StoplightViewProperties.CurrentColor);
			}
		}

		public string GreenDuration
		{
			get { return greenDurationTextBox.Text; }
			set
			{
				greenDurationTextBox.Text = value;
				RaisePropertyChanged(StoplightViewProperties.GreenDuration);
			}
		}

		public string YellowDuration
		{
			get { return yellowDurationTextBox.Text; }
			set
			{
				yellowDurationTextBox.Text = value;
				RaisePropertyChanged(StoplightViewProperties.YellowDuration);
			}
		}

		public string RedDuration
		{
			get { return redDurationTextBox.Text; }
			set
			{
				redDurationTextBox.Text = value;
				RaisePropertyChanged(StoplightViewProperties.RedDuration);
			}
		}

		public event EventHandler UpdateClicked;

		public event EventHandler ForceChangeClicked;

		public void SetError(string propertyName, string errorMessage)
		{
			Dictionary<string, Control> controlsByName = new Dictionary<string, Control>();
			controlsByName.Add(StoplightViewProperties.GreenDuration, greenDurationTextBox);
			controlsByName.Add(StoplightViewProperties.YellowDuration, yellowDurationTextBox);
			controlsByName.Add(StoplightViewProperties.RedDuration, redDurationTextBox);


			if(controlsByName.ContainsKey(propertyName))
			{
				errorProvider.SetError(controlsByName[propertyName], errorMessage);
			}
		}

		#endregion

		#region INotifyPropertyChanged Members

		public event PropertyChangedEventHandler PropertyChanged;

		#endregion

		// Event firing helpers

		protected virtual void RaisePropertyChanged(string propertyName)
		{
			PropertyChangedEventHandler handlers = PropertyChanged;
			if(handlers != null)
			{
				handlers(this, new PropertyChangedEventArgs(propertyName));
			}
		}

		protected virtual void RaiseUpdateClicked()
		{
			EventHandler handlers = UpdateClicked;
			if(handlers != null)
			{
				handlers(this, EventArgs.Empty);
			}
		}

		protected virtual void RaiseForceChangeClicked()
		{
			EventHandler handlers = ForceChangeClicked;
			if(handlers != null)
			{
				handlers(this, EventArgs.Empty);
			}
		}

		private void OnUpdateScheduleClicked(object sender, EventArgs e)
		{
			RaiseUpdateClicked();
		}

		private void OnChangeLightClicked(object sender, EventArgs e)
		{
			RaiseForceChangeClicked();
		}
	}
}
